src/pybind/mgr/affinity/module.py

```diff
@@ -0,0 +1,152 @@
+    COMMANDS = [
+        {
+            "cmd": "affinity create pool "
+                   "name=poolname,type=CephString,req=true "
+                   "name=region,type=CephString,req=true "
+                   "name=pd,type=CephString,req=true "
+                   "name=sd,type=CephString,req=true "
+                   "name=td,type=CephString,req=true",
+            "desc": "create a new pool with primary affinity to pd",
+            "perm": "rw"
+        },
+    ]
+
+    def __init__(self, *args, **kwargs):
+        super(Affinity, self).__init__(*args, **kwargs)
+
+        # set up some members to enable the serve() method and shutdown
+        self.run = True
+        self.event = Event()
+
+    def get_highest_rule_id(self):
+        crush = self.get('osd_map_crush')
+        crush_rules = crush['rules']
+        id = 1
+        for rule in crush_rules:
+            if rule['rule_id'] > id:
+                id = rule['rule_id']
+
+        return id
+
+    def create_crush_affined_rule(self, name, region, pd, sd, td):
+        rule_template = Template(""" rule ${name} {
+    id ${rule_id}
+    type replicated
+    min_size 1
+    max_size 10
+    step take ${region}${primary_domain}
+    step chooseleaf firstn 1 type host
+    step emit
+    step take ${region}${secondary_domain}
+    step chooseleaf firstn 1 type host
+    step emit
+    step take ${region}${tertiary_domain}
+    step chooseleaf firstn 1 type host
+    step emit
+}""")
+        
+        tempfd, crushmap_filename = mkstemp(dir='/tmp')
+
+        editable_crushmap_filename = crushmap_filename+".editable"
+
+        #get binary crush map
+        ret, buf, errs = self.rados.mon_command(json.dumps({"prefix":"osd getcrushmap"}), b'')
+        if ret != 0:
+            raise RuntimeError("Error getting binary crushmap. ret {}: '{}'".format(
+                ret, err))
+
+        # write binary crush map to file for crushtool
+        with open(crushmap_filename,"wb") as f:
+            f.write(bytearray(buf))
+
+        os.close(tempfd)
+
+        # decompile binary crush map
+        subprocess.check_call(["crushtool","-d",crushmap_filename,"-o",editable_crushmap_filename])
+
+        # write new rule to editable crush map, using template
+        with open(editable_crushmap_filename,"a") as cmfile:
+            cmfile.write(rule_template.render(
+                name=name, 
+                get_highest_rule_id()+1,
+                region=region,
+                primary_domain=pd,
+                secondary_domain=sd,
+                tertiary_domain=td
+            ))
+
+        # compile editable crush map
+        subprocess.check_call(["crushtool","-c",editable_crushmap_filename,"-o",crushmap_filename])
+
+        # read compiled, binary crush map
+        with open(crushmap_filename,"rb") as f:
+            barr = f.read()
+
+        # apply binary crush map
+        ret, buf, errs = self.rados.mon_command(json.dumps({"prefix":"osd setcrushmap"}), barr)
```

Comment:
there is an option ``prior_version`` argument to this command that you can pass the version you downloaded to in order to avoid race conditions (something else making a crush map update).  IIRC it is returned to stderr when you do getcrushmap.  If the versions don't match, you should loop and retry.

Reply:
TNX, I'll add that!